### PR TITLE
Feature: Add `@dateFormat` decorator

### LINKED ISSUE #282: Implement decorator to support date serialization format
- Implement 'dateFormat' decorator that allows choosing 'rfc1123', 'rfc3339', 'rfc7231', and 'unixTimeStamp' date formats
- Update emitters to respect the dateFormat set by the decorator
- Update emitters to set date format rfc7231 in http headers, and rfc3339 in other usages when no dateFormat is provided
- Update documentation to show usage of dateFormat decorator, and default date formats

Per design issue #https://github.com/Azure/cadl-azure/issues/1020

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #282

### CODE REVIEW

File: packages/compiler/lib/decorators.tsp [hunk 1]
```diff
@@ -337,6 +337,9 @@ extern dec projectedName(target: unknown, targetName: string, projectedName: str
  */
 extern dec discriminator(target: object | Union, propertyName: string);
 
+alias KnownDateFormat = "rfc1123" | "rfc7231" | "unixTimeStamp";
+extern dec dateFormat(target: zonedDateTime | ModelProperty, format: KnownDateFormat | string);
```
Comment (timotheeguerin) [inline packages/compiler/lib/decorators.tsp:341]: Does `@dateFormat` make the most sense as the name and not `@dateSerializationFormat`. It is different from `@format` on a string which actually dictate how the string is formatted

Comment (markcowl) [reply packages/compiler/lib/decorators.tsp:341]: I like shorter names, but dateSerializationFormat is more accurate

File: packages/openapi3/test/primitive-types.test.ts [hunk 2]
```diff
@@ -189,4 +189,56 @@ describe("openapi3: primitives", () => {
...
+      const res = await oapiForModel(
+        "MyDate",
+        `
+      scalar MyDate extends zonedDateTime;
+      `
+      );
+      deepStrictEqual(res.schemas.MyDate, { type: "string", format: "date-time" });
+    });
+
+    it("set format to 'timestamp' and type to 'ingeger' for 'unixtimestamp' format", async () => {
```
Comment (markcowl) [inline packages/openapi3/test/primitive-types.test.ts:204]:
```suggestion
    it("set format to 'timestamp' and type to 'integer' for 'unixtimestamp' format", async () => {
```

File: docs/standard-library/built-in-decorators.md [hunk 3]
```diff
@@ -531,6 +531,23 @@ model Dog extends Pet  {kind: "dog", bark: boolean}
 ```
 
 
+### `@dateFormat` {#@dateFormat}
+
+
+```typespec
+dec dateFormat(target: zonedDateTime | ModelProperty, format: rfc1123 | rfc7231 | unixTimeStamp | string)
```
Comment (markcowl) [inline docs/standard-library/built-in-decorators.md:538]: The values should be rfc3339 and rfc7231.  rfc7231 is the replacement for rfc1123

File: packages/compiler/lib/decorators.ts [hunk 4]
```diff
@@ -533,6 +533,31 @@ export function isSecret(program: Program, target: Type): boolean | undefined {
   return program.stateMap(secretTypesKey).get(target);
 }
 
+// -- @dateFormat decorator ---------------------
+export type KnownDateFormat = "rfc1123" | "rfc7231" | "unixTimeStamp";
```
Comment (markcowl) [inline packages/compiler/lib/decorators.ts:537]:
```suggestion
export type KnownDateFormat = "rfc3339" | "rfc7231" | "unixTimeStamp";
```

File: packages/openapi3/src/openapi.ts [hunk 5]
```diff
@@ -1307,6 +1308,16 @@ function createOAPIEmitter(program: Program, options: ResolvedOpenAPI3EmitterOpt
       newTarget.format = "password";
     }
 
+    const dateFormat = getDateFormat(program, typespecType);
```
Comment (markcowl) [inline packages/openapi3/src/openapi.ts:1311]: We should also set the default dateFormat, which requires knowing whether the date is used in a header  (rfc7231) or elsewhere (rfc 3339)

File: packages/openapi3/test/primitive-types.test.ts [hunk 6]
```diff
@@ -189,4 +189,56 @@ describe("openapi3: primitives", () => {
       });
     });
   });
+
+  describe("date format", () => {
+    it("set format to 'date-time' by default", async () => {
```
Comment (markcowl) [inline packages/openapi3/test/primitive-types.test.ts:194]: Should have an alternate where this is ued in a header, and the default format is rfc-7231

**Review by markcowl — CHANGES_REQUESTED** [review]

**Comment by bterlson** [comment]
Sorry forgot the context, remind me why this isn't just the existing format decorator?

Also how does this combine with utcDateTime and offsetDateTime? What are the semantic differences between say a utcDateTime without a format and an offsetDateTime with a unixTimestamp format?

File: packages/openapi3/src/openapi.ts (hunk 5 above)
Comment (timotheeguerin) [reply packages/openapi3/src/openapi.ts:1311]: We’ll in openapi date-time already assume that from what I understand.

**Comment by timotheeguerin** [comment]
@bterlson Think that design discussion was done when I was one leave so can’t say about what was discussed but from my comment above I think it is different from format as it means serialization format vs format means a know pattern for the string.

for the utc and offset date time I am not fully sure either what was the rationale behind that distinction

**Comment by bterlson** [comment]
I was probably present for the discussion but I've forgotten it. Forgive me! (Also I've spent a bunch of time in Json Schema land and paging back in arcane date knowledge)

I worry that this decorator's design isn't fully rationalized with existing/planned features so it might be worth circling back a little? Some thoughts:

The existing format decorator isn't just for string types (despite what the docs say, int16 is an example integer format). Json Schema leaves format unbounded and allows it to be used for any type. We should probably too? (Or the json schema emitter will need yet another format decorator?)

The Json Schema and OAS include serialization format in its format registry - format: "date-time" requires RFC3339. Seems reasonable to say that we're extending this list with some additional known formats. A somewhat orthogonal axis might make it hard for OAS and JsonSchema emitters to know what combinations are valid/when to emit a registered format vs. fall back to pattern or something?

The idea behind utcDateTime and offsetDateTime is to describe the contract the service is exposing with regard to whether it expects/persists/round trips time zones and offsets (zonedDateTime), offsets (offsetDateTime) or just ticks since epoch in UTC (utcDateTime).

Format is mostly orthogonal to this notion, except that some formats might imply information loss depending on type it's applied to. I think this should be an error probably? `unixTimestamp` should not be applied to either offsetDateTime or utcDateTime - if you want a unixTimestamp, you should use `utcDateTime` because you're saying clients will not send and your service will not accept or persist either offsets or time zones.

**Comment by timotheeguerin** [comment]
My argument for not using format still applies. In json schema we are saying the value is a string with a specific format (date time). Or an int being something more specific maybe.

but here we have already that distinction of a type which is where it feels weird to me 

- format(in jsonschema and openapi), @format -> precise the type
- Date format(or what we want to do here) -> specify how we can serialize this type of the wire 

maybe this is not the right solution but format to me sounds like the wrong one as well

**Comment by bterlson** [comment]
Doesn't format: date-time in OpenAPI/Json Schema describe the serialization format moreso than the data type? date-time requires RFC3339, meanwhile format is type-agnostic, the type is described by the `type` field (though date-time only makes sense on a string).

**Review by daviwil — APPROVED** [review]

**Comment by timotheeguerin** [comment]
@bterlson I guess it does technically but to me it still sounds like `@format` is mean to represent a certain pattern for the data, which in the case of jsonschema and openapi is a Serialization format.

Like if we look at those format in jsonschema:
- uuid:  string is representing a uuid
- uri: string is representing a uri
- date-time: string is representing a rfc3339 date time
- int8: that integer can only be from -127 to 128

Using format on the date times would to me imply similar style of pattern restriction

Like, not that it make sense but if you said `@format("date")` on `utcDateTime`) to specify that even though this is using a full date time representation, the time will always be 0.

File: packages/openapi3/src/openapi.ts (hunk 5 above)
Comment (markcowl) [reply packages/openapi3/src/openapi.ts:1311]: AFAIK, no, see: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md?plain=1#L170

date-time should be the default for non-header types, but for headers, we need to be explicit

Comment (timotheeguerin) [reply packages/openapi3/src/openapi.ts:1311]: Oh I see, I’ll update

**Comment by bterlson** [comment]
In a world where pattern can apply to any type, isn't serialization format just a pattern restriction? I see your point if we say `format` is basically `stringFormat` as effectively a well-known list of regexps for `pattern`, but in a world where `format` can apply to any type, I don't know that the distinction is meaningful. At the least, it seems hard to describe to people when one is used over the other.

**Comment by timotheeguerin** [comment]
I guess an argument against that decorator is, let’s say we have this new geolocation type, I guess that also needs to specify how to serialize it. Seems overkill to have to have a new decorator per month primitive type that might serialize in many different ways

**Comment by markcowl** [comment]
`@format` -> type format in openapi
`@serializationFormat` -> general formatting hints that might encompass openapi format and other formatting mechanisms

Note that, if we have a core decorator that serves the purpose of `@format`, then we likely want to call it something else
-or -

`@format`

-or -

type-specific format decorators

// we decided against type-specific decorators

**Comment by timotheeguerin** [comment]
# `@format` vs `@serializeTo` decorators

Description:

- `@format`: Existing decorator meant to provide an known pattern for the type
- `@serializeTo` (or `@serializationFormat`) Is meant to provide the serialization protocol for that type over the wire

## Examples

### Specify uuid, email, ipv4, ipv6

```tsp
model Bar {
  @format("uuid") id: string;
  @format("email") userEmail: string;
  @format("ipv4") ipv4: string;
  @format("ipv6") ipv6: string;
}
```

### Specify rfc3339

```tsp
model Bar {
  @serializeTo("rfc3339")
  createdAt: utcDateTime;

  @serializeTo("rfc3339")
  modifiedAt: offsetDateTime;

  @serializeTo("rfc3339")
  runtime: duration;

  // We could also technically represent the datetime as a string and then we'd use the `@format` decorator to specify the pattern of that string
  @format("date-time")
  createdAtButString: string;
}
```

### Specify rfc7231

```tsp
model Bar {
  @serializeTo("rfc7231")
  createdAt: utcDateTime;

  @serializeTo("rfc7231")
  modifiedAt: offsetDateTime;

  @serializeTo("rfc7231")
  runtime: duration;

  @format("date-time-rfc7231")
  createdAtButString: string;
}
```

### Specify unixTimestamp

```tsp
model Bar {
  @serializeTo("unixTimestamp")
  createdAt: utcDateTime;

  @serializeTo("unixTimestamp")  // invalid
  modifiedAt: offsetDateTime;

  @serializeTo("unixTimestamp") // invalid
  runtime: duration;


}
```

### Specify Duration serialized as seconds

```tsp
model Bar {
  @serializeTo("seconds-int")
  runtimeAsDouble: duration;
  @serializeTo("seconds-float64")
  runtimeAsInt: duration;
}
```

Alternative here could be that `@serializeTo` takes a 2nd optional parameter that specify the type it serialize to. By default being a string

```tsp
model Bar {
  @serializeTo("seconds", int32)
  runtimeAsDouble: duration;
  @serializeTo("seconds", float64)
  runtimeAsInt: duration;
}
```

**Comment by timotheeguerin** [comment]
Replaced by updated `@encode` decorator https://github.com/microsoft/typespec/pull/1899

**Pull request closed without merging**